### PR TITLE
quiche: early fail listener config with both quic and connection_balencer

### LINKED ISSUE #13116: quic: disallow ConnectionBalancer config
ConnectionBalancer isn't currently supported for udp/quic, and implementing it would probably be CPU-inefficient due to packet forwarding between workers. Disallow this config.

This should be done before MVP as it would other invalidate a previously-loadable config.

### PR DESCRIPTION
**Pull request by danzh2010**
Signed-off-by: Dan Zhang <danzh@google.com>

QUIC listener use BPF kernel routine for each packet to guarantee stable connection. [connection_balencer](https://github.com/envoyproxy/envoy/blob/ba474ac375418d5529a30a9510a8ff5f0a5ada9a/generated_api_shadow/envoy/config/listener/v3/listener.proto#L256) in Listener config undermines this purpose. So it shouldn't be used with QUIC listener.

Risk Level: low
Testing: added unit test
Fixes #13116

### CODE REVIEW

File: source/server/listener_impl.cc [hunk 1]
```diff
@@ -444,6 +444,10 @@ void ListenerImpl::buildUdpListenerFactory(Network::Socket::Type socket_type,
   udp_listener_config_ = std::make_unique<UdpListenerConfigImpl>(config_.udp_listener_config());
   if (config_.udp_listener_config().has_quic_options()) {
 #ifdef ENVOY_ENABLE_QUIC
+    if (config_.has_connection_balance_config()) {
+      throw EnvoyException("connection_balance_config is configured for QUIC listener which "
+                           "doesn't work with connection balencer.");
```
Comment (ggreenway) [inline source/server/listener_impl.cc:449]: spelling: balancer

**Review by ggreenway — CHANGES_REQUESTED** [review]
Thanks for fixing this! Looks good except for one typo.

/wait

File: source/server/listener_impl.cc (hunk 1 above)
Comment (danzh2010) [reply source/server/listener_impl.cc]: fixed

**Pull request merged**